### ISSUE TITLE
PoDoFo documentation ?

### ISSUE THREAD

**Issue by seba1204** [podofo/podofo#91]
TL;DR
**I don't find any documentation for this lib.**

Hello, thanks for this amazing lib ! I'd like to use it, but I don't know how. I'm probably blind, but I couldn't find any documentation.
**_Maybe a simple link to the README.md file could be added, even from an old version?_**

I've found some links:
- [podofo.github.io](https://podofo.github.io/)
- [sourceforge.net/projects/podofo/](https://sourceforge.net/projects/podofo/)
- [github.com/podofo/podofo](https://github.com/podofo/podofo)
- [podofo.sourceforge.net/doc/html/classPoDoFo_1_1PdfMemDocument.html](http://podofo.sourceforge.net/doc/html/classPoDoFo_1_1PdfMemDocument.html#a0b37e7cac8642bb4e7190d5c559f8603)
- [wikipedia.org/wiki/Podofo](https://en.wikipedia.org/wiki/Podofo)

But none of them seems to be a real detailed documentation.
I seem to remember a few months ago coming across more details, but I can't find anything.

Could you help me?
Thanks!

**Comment by Nadwey** [comment]
I also couldn't find any docs, so I setup a repo that generates them every week

https://nadwey.github.io/podofo-docs/

**Comment by ceztko** [comment]
I'm sorry, but I'm not the right person to write beginners focused documentation: 100% of the effort is now put on creating a very powerful API that is high level enough that people won't need to bother too much with PDF internals, but it's a very big task and I can't guarantee I am succeeding in every aspect of this big work. I made all I could to keep updated the Doxygen documentation, but for sure I'm forgetting a lot methods, especially the many overloads added for which I would like to not clutter the headers too much. To me, Doxygen documentation duplication smells as bad as code duplication because it also gets outdated very quickly: suggestions welcome on how to handle this.

I never generated the documentation myself so I would be glad to integrate a github [workflow](https://github.com/podofo/podofo/tree/master/.github/workflows) that:
- Generate the documentation
- Upload it to a github repository. In the end I will push it to github pages repository so I can put a link to it.

Patch welcome.

**Comment by ferencIOS** [comment]
Hello,

I've made an initial attempt to address this issue. In this branch: [feature/issue-91](https://github.com/podofo/podofo/compare/master...ferencIOS:podofo:feature/issue-91), I've tried to create automated documentation using Doxygen. You can download and see the results [here](https://github.com/podofo/podofo/files/13214443/documentation.zip).

Before you delve into it, here are some considerations regarding this attempt:
- **Platform Testing**: The entire testing was done on `macOS`.
- **Workflow Limitations**: The documentation workflow is currently missing for `Linux` and `Windows`.
- **Nature of Work**: Please note that this is merely a **Proof of Concept (POC)**, and the idea is to extract specific tasks from this for more comprehensive work.

I hope this helps, and I'm keen on receiving any feedback or suggestions you might have on the documentation. Thank you!

**Comment by ceztko** [comment]
Hello @ferencIOS. Thank you for your efforts. Some comments:

> Workflow Limitations

I believe there's no need to have workflow steps in each platform. One separate workflow doing just documentation would be good enough, which should be possible I believe. So I recommend to strip the documentation part from the mac workflow and leave that unmodified.

> Platform Testing

I would prefer the workflow to run on linux, since it's just faster on github (mac workflows are very slow). That's something I can port later, just in case.

I also would like to see the github git repository push step in action.

**Comment by ferencIOS** [comment]
Hello @ceztko 

Thank you for your feedback. Following your suggestions, I have refined the workflow to align with the project's requirements:

1. **Workflow Streamlining:** The documentation steps have been isolated from the macOS workflow. There is now a separate workflow solely dedicated to documentation generation.

2. **Platform Choice:** Although the initial tests were conducted on macOS, I concur with your preference for Linux due to its quicker execution on GitHub. The current workflow now utilizes `ubuntu-latest` for documentation generation.

3. **Documentation Deployment:** I've successfully implemented the GitHub repository push step, and the automated documentation is now live. You can view the deployed documentation at [https://ferencios.github.io/podofo/](https://ferencios.github.io/podofo/).

The associated branch and the workflow file can be reviewed here:
[https://github.com/podofo/podofo/compare/master...ferencIOS:podofo:feature/issue-91](https://github.com/podofo/podofo/compare/master...ferencIOS:podofo:feature/issue-91)

Please note the following workflow details:

```yaml
name: build-documentation

on:
  push:
    branches: [appropriate-branch]
  pull_request:
    branches: [appropriate-branch]
  
  workflow_dispatch:

permissions:
  contents: read
  pages: write
  id-token: write

concurrency:
  group: "pages"
  cancel-in-progress: false

env:
  BUILD_TYPE: Release

jobs:
  build-and-deploy:
    runs-on: ubuntu-latest
    steps:
      - name: Checkout code
        uses: actions/checkout@v4
      # Other steps truncated for brevity...
  deploy:
    needs: build-and-deploy
    runs-on: ubuntu-latest
    environment:
      name: github-pages
      url: ${{ steps.deployment.outputs.page_url }}
    steps:
      - name: Deploy to GitHub Pages
        id: deployment
        uses: actions/deploy-pages@v2
```
In alignment with the environment protection rules, I am ready to transition the workflow to the appropriate branch that is permitted to deploy to `github-pages`.

To ensure the workflow functions correctly and complies with the repository's environment protection rules, the following steps have been taken:
* Navigate to the GitHub repository settings.
* Access the "Environments" section.
* Select the github-pages environment to review the protection rules that are in place.

**Comment by ceztko** [comment]
@ferencIOS Please, can you squash and create a PR? I will handle merging with my customization later.

**Comment by ceztko** [comment]
> To ensure the workflow functions correctly and complies with the repository's environment protection rules, the following steps have been taken:
> 
> * Navigate to the GitHub repository settings.
> * Access the "Environments" section.
> * Select the github-pages environment to review the protection rules that are in place.

These steps are not clear to me. I can create a `github-pages` environment, but then what I should do?

Also I'm not entirely sure to which repository the pages will be uploaded to and at which location. Maybe it's set automatically to push the default `<repository>.github.io` repo, but I'm not entirely sure. Also I would like the location to be https://podofo.github.io/documentation and not https://podofo.github.io/podofo, as it appears to be in your environment.

**Comment by ceztko** [comment]
Hi @ferencIOS . Again, if you can create a PR (as is, or rebased and squashed), otherwise it's a no start.

**Comment by ferencIOS** [comment]
Hi @ceztko!

The PR can be found here: https://github.com/podofo/podofo/pull/120.

Now, it is necessary for you to perform a few steps in your GitHub repository settings.
Here's a brief guide:

- Navigate to the 'Settings' of your GitHub repository.
- Access the 'Environments' section within the settings.
- Select the 'github-pages' environment.
- Review and adjust the protection rules as needed to accommodate the new documentation.

Let me know if I can help you.

**Comment by ceztko** [comment]
Hi @ferencIOS . These steps are not clear at all for me:

> - Select the 'github-pages' environment.
> - Review and adjust the protection rules as needed to accommodate the new documentation.

Let me show what I have in my panels. If I go to Ennviroments, there was no pre-created `github-pages` environment, I had to create it myself with the "New environment" button.

![image](https://github.com/podofo/podofo/assets/3037449/c5c59b39-6803-49b6-be30-673cef8fa0ea)

Then if I enter such created environment that's what I see:

![image](https://github.com/podofo/podofo/assets/3037449/4da7cb0d-130e-404a-8128-049bd6444ae6)

In other words there's nothing obvious to do as when you say "Review and adjust the protection rules as needed to accommodate the new documentation.". Please assume I know nothing about "protection rules". You will have to go in a more detailed explanation, or show me a screenshot of your settings.

**Comment by ceztko** [comment]
Ok, I did some progress. You should really have specified that I must go to the [podofo.github.io](https://github.com/podofo/podofo.github.io/) settings, and not the podofo main repository. Here I actually found a `githu-pages` environment already created. In the configuration panel that's what I see (below). Again, for me it's not really clear what to do here. Assume you have to tell your mother :) ...

![image](https://github.com/podofo/podofo/assets/3037449/445a67d5-365b-4fed-b5cc-c55232da0477)

**Comment by ferencIOS** [comment]
> Let me show what I have in my panels. If I go to Ennviroments, there was no pre-created github-pages environment, I had to create it myself with the "New environment" button.

1)
Yes, you have to do it.
<img width="828" alt="Screenshot 2023-12-19 at 23 06 26" src="https://github.com/podofo/podofo/assets/39296224/1eea5ea7-7278-40ec-884e-1f5a7f085d72">

2) 
<img width="826" alt="Screenshot 2023-12-19 at 23 06 50" src="https://github.com/podofo/podofo/assets/39296224/e401a3e9-1340-46f2-885d-21335faa6bfa">

3)
<img width="817" alt="Screenshot 2023-12-19 at 23 07 11" src="https://github.com/podofo/podofo/assets/39296224/297e37eb-4cce-4641-8564-2dd4af2ee133">

Please switch to the most appropriate branch: `main` is ok

4)
update the YAML configuration for triggering actions on chosen branch

`.github/workflows/build-documentation.yml`
```yml
push:
  branches: [chosen-branch-name]
pull_request:
  branches: [chosen-branch-name]
```

<img width="1125" alt="Screenshot 2023-12-19 at 23 15 43" src="https://github.com/podofo/podofo/assets/39296224/d0f475ea-066e-4738-b6a7-c57b0b9dd994">

**Comment by ceztko** [comment]
Ok, somehow I think the `github-pages` environment wasn't really necessary, and as a matter of fact I deleted it in the [podofo](https://github.com/podofo/podofo) repository. Actually the pages are not committed to the [podofo.github.io](https://github.com/podofo/podofo.github.io) repository as I expected, but they are still visible in a "virtual" http location called `podofo` within the project pages entry point https://podofo.github.io. I changed the html output directory to be called `documentation` instead so, the final link for the documentation is https://podofo.github.io/podofo/documentation/ . Actually one step that was necessary to have the doxygen workflow was enabling deploy through Github actions.in the `Pages` tab of the [podofo](https://github.com/podofo/podofo) repository (screenshot below), otherwise I had an [error](https://github.com/podofo/podofo/actions/runs/7276736773/job/19827182200). More changes are planned, as create a target for doxygen generation within CMake, but for now this is done. Thank you @ferencIOS !

![image](https://github.com/podofo/podofo/assets/3037449/0e6eb0e1-ae8a-45fc-a85c-a5cec040b8d7)

**Comment by ceztko** [comment]
I updated my last comment with more discoveries: eventually the `github-pages` was created automatically by enabling Pages deploy through actions. No need to perform any changes in it, though.
